middleware/auth_bearer.go

```diff
@@ -36,30 +37,86 @@ func BearerAuthFuncGRPC(auth authentication.Authentication) grpc_auth.AuthFunc {
 		if err != nil {
 			return nil, status.Error(codes.Unauthenticated, err.Error())
 		}
-		return injectClaims(ctx, token)
+		return claimInjector(ctx, token)
+	}
+}
+
+// ClaimInjector allows authentication layers to inject JWT claims into a context.Context.
+type ClaimInjector func(ctx context.Context, token jwt.Claims) (context.Context, error)
+
+// ClaimInjectorBehavior is used in combination with ClaimInjector when grouping
+// different claim injectors by using GroupClaimInjectors.
+type ClaimInjectorBehavior func(ctx context.Context, err error) (context.Context, error)
+
+// GroupClaimInjectors treats all the given injectors as a single one.
+//
+// By setting a MandatoryInjection, the resulting injector will early return
+// if an error is found at any point during the claim injection.
+//
+// If OptionalInjection is used instead, no errors will be returned during the
+// claim injection.
+//
+// If no ClaimInjectorBehavior is provided, it defaults to MandatoryInjection.
+//
+// Example:
+//
+//	GroupClaimInjectors(MandatoryInjection,
+//		GroupClaimInjectors(MandatoryInjection, SubjectClaimer),
+//		GroupClaimInjectors(OptionalInjection, EmailClaimer),
+//	)
+func GroupClaimInjectors(behavior ClaimInjectorBehavior, injectors ...ClaimInjector) ClaimInjector {
```

Comment:
I find it more ergonomic to flip the way the behavior is defined. Instead of having a function that receives the behavior and a set of injectors, have a function for each behavior and handle the complexity internally.

```go
// Make the existing injector functions private
func groupClaimInjectors(behavior ClaimInjectorBehavior, injectors ...ClaimInjector) ClaimInjector { [...] }
func mandatoryInjection(ctx context.Context, err error) (context.Context, error) { [...] }
func optionalInjection(ctx context.Context, err error) (context.Context, error) { [...] }

func GroupMandatoryClaimInjectors(injectors ...ClaimInjector) ClaimInjector {
	groupClaimInjectors(MandatoryInjection, injectors)
}

func GroupOptionalClaimInjectors(injectors ...ClaimInjector) ClaimInjector {
	groupClaimInjectors(OptionalInjection, injectors)
}
```

This also makes it more consistent with the Unary and Stream interceptors that are used by gRPC in close proximity to these injectors.

Not a strong opinion. Feel free to resolve if you disagree.